examples/advanced/visual_sudoku.py

```diff
@@ -0,0 +1,281 @@
+        vizsudoku[val_idx] = torch.stack([testset[i][0] for i in idx])
+
+    return vizsudoku
+
+
+
+from torch import nn 
+import torch.nn.functional as F
+
+class LeNet(nn.Module):
+    '''
+        TODO: Write doc+ reference to paper
+    '''
+    def __init__(self):
+        super(LeNet, self).__init__()
+        self.conv1 = nn.Conv2d(1, 6, 5, 1, padding=2)
+        self.conv2 = nn.Conv2d(6, 16, 5, 1)
+        self.fc1 = nn.Linear(5*5*16, 120)
+        self.fc2 = nn.Linear(120, 84)
+        self.fc3 = nn.Linear(84,10)
+
+    def forward(self, x):
+        x = F.relu(self.conv1(x))
+        x = F.max_pool2d(x, 2, 2)
+        x = F.relu(self.conv2(x))
+        x = F.max_pool2d(x, 2, 2)
+        x = x.view(-1, 5*5*16) 
+        x = F.relu(self.fc1(x))
+        x = self.fc2(x)
+        x = self.fc3(x)
+        return F.log_softmax(x, dim=1)
+
+def load_clf(clf_classname, path):
+    '''
+        TODO: Write doc.
+    '''
+    net = clf_classname()
+    state_dict = torch.load(path, map_location=lambda storage, loc: storage)
+    net.load_state_dict(state_dict)
+    return net
+
+
+@torch.no_grad() #TODO: What does this do ?
+def predict_proba_sudoku(model, vizsudoku):
+    #TODO: Change shape of output to work for differnet kinds of sudoku shapes (4x4, 9x9, ...)
+    # reshape from 9x9x1x28x28 to 81x1x28x28
+    pred = model(vizsudoku.flatten(0,1))
+    # our NN returns 81 probabilistic vector: an 81x10 matrix
+    return pred.reshape(9,9,10).detach() # reshape as 9x9x10 tensor for easier visualisation
+
+if __name__ == '__main__':
+    '''
+        Missing a bit of explanations even small bit
+    '''
+    e = 0 # empty cell
+    sample_sudoku_grid = np.array([
+        [e, e, e,  2, e, 5,  e, e, e],
+        [e, 9, e,  e, e, e,  7, 3, e],
+        [e, e, 2,  e, e, 9,  e, 6, e],
+
+        [2, e, e,  e, e, e,  4, e, 9],
+        [e, e, e,  e, 7, e,  e, e, e],
+        [6, e, 9,  e, e, e,  e, e, 1],
+
+        [e, 8, e,  4, e, e,  1, e, e],
+        [e, 6, 3,  e, e, e,  e, 8, e],
+        [e, e, e,  6, e, 8,  e, e, e]
+    ])
+
+    # Emilio: Missing a bit of context here ?
+    vs = sample_visual_sudoku(sample_sudoku_grid)
+
+    # TODO: (Minor) Does not work from outside of folder + what is clf ? (classifier?)
+    model = load_clf(LeNet,'lenet_mnist_e15.pt')
+
+    # (log)probabilities for each cell
+    logprobs = predict_proba_sudoku(model, vs)
+
+    # maximum likelihood class 
+    ml_digits = np.argmax(logprobs, axis=-1)
+
+    # Emilio:  dvar ? decision vars (puzzle_vars ?)
+    dvar, sudoku_model = model_sudoku(sample_sudoku_grid)
+
+    # Emilio: sol3 ? where is sol 1 and 2?
+    sol3 = solve_vizsudoku_hybrid2(dvar, sudoku_model,logprobs, sample_sudoku_grid)
+
+    sudoku_model += [dvar[sample_sudoku_grid > 0] == sample_sudoku_grid[sample_sudoku_grid > 0]]
+
+    # TODO: why do you need to create a new model ?
+    # old: model = Model(cons)
+
+    model.solve()
+    solution = dvar.value()
+
+    # TODO: pretty printing ?
```

Comment:
work on the flow...

You first want to compute the argmax solution, so have only the code for that, and print that result.
Some visu/print/annotation of how it makes a mistake would be good.... (a notebook could be better then? also forces you to be more linear...)

Then, write that you will now make a max likelihood formulation.

Ideally, the reader can see the difference very close to each other, e.g.

m = base_sudoku(grid)
m += (grid[given] == argmax[given])
m.solve()

m2 = base_sudoku(grid)
m2.optimize(np.sum((grid==i)*logprobs[i] for i in range()) # or some fancy one-liner to write this ; )
m2.solve()


THAT would be a great example...

then, advanced edition, can build on m2 perhaps?
while not_unique(grid.value()):
  m2 += # blocking clause
  m2.solve()


I'm not sure if it is doable, but if your 'def visu_sudoku():` would contain this, it would be a very pedagogical example...